### PR TITLE
Develop upstream sync 190520

### PR DESCRIPTION
**Pull request by sunway513**
Merge conflicts list:      
        both modified:   tensorflow/compiler/xla/service/gpu/BUILD
        both modified:   tensorflow/compiler/xla/service/gpu/cudnn_conv_runner.cc
        both modified:   tensorflow/contrib/reduce_slice_ops/kernels/reduce_slice_ops_gpu.cu.cc
        both modified:   tensorflow/contrib/resampler/kernels/resampler_ops_gpu.cu.cc
        both modified:   tensorflow/contrib/seq2seq/kernels/beam_search_ops_gpu.cu.cc
        both modified:   tensorflow/core/kernels/BUILD
        both modified:   tensorflow/core/kernels/check_numerics_op_gpu.cu.cc
        both modified:   tensorflow/core/kernels/concat_lib_gpu_impl.cu.cc
        both modified:   tensorflow/core/kernels/cwise_ops.h
        both modified:   tensorflow/core/kernels/fused_batch_norm_op.cc
        both modified:   tensorflow/core/kernels/lrn_op.cc
        both modified:   tensorflow/core/kernels/reduction_gpu_kernels.cu.h
        both modified:   tensorflow/core/kernels/svd_op_gpu.cu.cc
        both modified:   tensorflow/core/kernels/tridiagonal_matmul_op_gpu.cu.cc
        both modified:   tensorflow/core/util/gpu_device_functions.h
        both modified:   tensorflow/core/util/gpu_kernel_helper.h
        both modified:   tensorflow/core/util/gpu_kernel_helper_test.cu.cc
        both modified:   tensorflow/python/distribute/BUILD
        both modified:   tensorflow/python/keras/distribute/BUILD
        both modified:   tensorflow/stream_executor/gpu/gpu_driver.h
        both modified:   tensorflow/tools/ci_build/Dockerfile.gpu

### CODE REVIEW

**Comment by sunway513** [comment]
Here's NV path failure info, @deven-amd could you help review it? 
```
[31m[1mERROR: [0m/workspace/tensorflow/compiler/jit/BUILD:71:12: Traceback (most recent call last):
	File "/workspace/tensorflow/compiler/jit/BUILD", line 68
		cc_library(name = "xla_gpu_jit", visibility =..."], <2 more arguments>)
	File "/workspace/tensorflow/compiler/jit/BUILD", line 71, in cc_library
		if_cuda_is_configured(if_cuda([":jit_compilation_passe..."])) + if_rocm_is_configured(if_rocm([":jit_compilation_passe..."]))
'+' operator applied to incompatible types (select of select, list)

**Comment by deven-amd** [comment]
will look into this later tonight

**Comment by deven-amd** [comment]
this commit is what is causing us grief:

https://github.com/tensorflow/tensorflow/commit/36e4025cad1fd463e5dbc9d53daf65af8e1547ef

@whchung we need to figure out G's intent behind this change...if we need to make a similar change to the rocm equivalent, it will most likely cause some hiccups.

The quickest "fix" for us maybe to simply undo this change...I will try that locally and if that works I will push out the commit

**Comment by whchung** [comment]
Return select instead of a list would mean some amount of change on the bazel script. For the time being let’s revert this change in our downstream fork.

**Comment by deven-amd** [comment]
yeah...reverting that change seems to get past the initial error, but am running into other issues now...working through them...will push out commits with my progress before I sign off for the day

**Comment by deven-amd** [comment]
pushed out some commits that will make the non-xla `cuda` path to compile.

the XLA `cuda` compile still has errors

**Comment by whchung** [comment]
some quick observation:

- https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream-sync-190520/tensorflow/core/kernels/avgpooling_op.cc is again incorrectly not carrying `TENSORFLOW_USE_ROCM` macro which caused quite a few pooling-related tests to fail on `rocm` and `rocm-v2` paths.
- https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream-sync-190520/tensorflow/core/kernels/cwise_ops.h#L1015 , `std::atan2()` is not available on ROCm on GPU path, may need to use `::atan2()` or `::atan2f()`. check HIP documentation.

- `cuda` path seems to have suffer from interface change

File: tensorflow/core/kernels/cwise_ops.h [hunk 1]
```diff
@@ -1012,7 +1012,7 @@ struct scalar_atan2_op {
   EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE Scalar
   operator()(const Scalar& y, const Scalar& x) const {
 #if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
-    return std::atan2(y, x);
+    return ::atan2(y, x);
```
Comment (whchung) [inline tensorflow/core/kernels/cwise_ops.h]: `GOOGLE_CUDA` path allows `std::atan2` so here we might need to change to something like:

```
#if GOOGLE_CUDA
return std::atan2(y, x);
#elif TENSORFLOW_USE_ROCM
return ::atan2(y, x);
#else
return std::atan2(y, x);
#endif
```

Comment (sunway513) [reply tensorflow/core/kernels/cwise_ops.h]: fixed in 3efa5f4

**Comment by sunway513** [comment]
rocm CI has three failed cases:
- //tensorflow/python/keras:integration_test
this test passed in my local environment
- //tensorflow/python/distribute:keras_saved_model_test
- //tensorflow/python/distribute:keras_saved_model_test_gpu 
those two tests are newly added, failed in eager mode.

**Comment by sunway513** [comment]
NV path got the following build error in my local environment:
```In file included from /usr/lib/gcc/x86_64-linux-gnu/5/include/immintrin.h:59:0,
                 from external/eigen_archive/Eigen/src/Core/util/ConfigureVectorization.h:318,
                 from external/eigen_archive/Eigen/Core:22,
                 from ./third_party/eigen3/Eigen/Core:1,
                 from ./tensorflow/compiler/xla/service/cpu/runtime_single_threaded_conv2d.h:19,
                 from tensorflow/compiler/xla/service/cpu/runtime_single_threaded_conv2d.cc:16:
/usr/lib/gcc/x86_64-linux-gnu/5/include/avx512vlbwintrin.h:105:1: error: inlining failed in call to always_inline ‘__m256i _mm256_maskz_loadu_epi16(__mmask16, const void*)’: target specific option mismatch
 _mm256_maskz_loadu_epi16 (__mmask16 __U, void const *__P)_

**Comment by sunway513** [comment]
@deven-amd  @whchung , do you have any ideas on the NV build failure?
The command failed while compiling `tensorflow/compiler/xla/service/cpu/runtime_single_threaded_conv2d.cc`, however, that file has not been changed for a couple of months. No moves on Eigen commit as well.

**Comment by whchung** [comment]
Check if Eigen pointers agree. Eigen commit was moved.
this PR: https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream-sync-190520/tensorflow/workspace.bzl#L153
upstream: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/workspace.bzl#L152

**Comment by deven-amd** [comment]
was trying to reproduce this error locally (by trying to do a XLA build) and I get the following error
```
...
tensorflow/compiler/xla/service/gpu/nvptx_compiler.cc: In member function 'std::vector<unsigned char, std::allocator<unsigned char> > xla::gpu::NVPTXCompiler::CompilePtxOrGetCachedResult(stream_executor::StreamExecutor*, const string&, int, int, const xla::HloModuleConfig&)':
tensorflow/compiler/xla/service/gpu/nvptx_compiler.cc:683:48: error: 'PtxOptsFromConfig' was not declared in this scope
             PtxOptsFromConfig(hlo_module_config));
...
```

m assuming there are two issues at play here, one above and the other as reported by @sunway513 .

let me work on the one I am running into

**Comment by sunway513** [comment]
@whchung the eigen pointer moved after I started the weekly merge process on 190520. But it seems they've revoked the eigen pointer again right after.
Let me try different eigen pointers and see if it is feasible for us to move it without rebasing against the upstream master branch.

**Comment by sunway513** [comment]
12 rocm-xla path tests were failed with the following signature:
`2019-05-23 23:01:27.190143: I tensorflow/stream_executor/stream.cc:5499] [stream=0x807bba0,impl=0x7f08cc0] Internal: stream did not block host until done; was already in an error state`
`2019-05-23 23:01:27.190195: W tensorflow/stream_executor/stream.cc:274] Error blocking host until done in stream destructor: Internal: stream did not block host until done; was already in an error state`

**Comment by whchung** [comment]
@parallelo to give you an heads up. I'm working on getting this PR merged. please hold the weekly merge for this week until this PR gets merged. I'll keep you posted.

**Comment by parallelo** [comment]
@whchung - Sounds good.  Thx for the heads-up.

**Comment by whchung** [comment]
@parallelo I've fixed the issue on `rocm-xla` path. Running tests on CI now. I believe the PR is in good shape now. Once this PR is merged you can start the process for this week.

**Comment by parallelo** [comment]
Awesome.  I'll watch for the merge event, then will push forward with this week's iteration.

**Pull request merged**